memflow-derive/src/lib.rs

```diff
@@ -69,91 +40,142 @@ pub fn connector(args: TokenStream, input: TokenStream) -> TokenStream {
+                out: &mut ::memflow::plugins::connector::MUConnectorInstance
+            ) -> i32 {
+                ::memflow::plugins::connector::create_without_logging(args, out, #func_name)
             }
         }
     };
 
-    let mut gen = quote! {
+    let gen = quote! {
         #[doc(hidden)]
         #[no_mangle]
         pub static MEMFLOW_CONNECTOR: ::memflow::plugins::ConnectorDescriptor = ::memflow::plugins::ConnectorDescriptor {
-            connector_version: ::memflow::plugins::MEMFLOW_CONNECTOR_VERSION,
+            connector_version: ::memflow::plugins::MEMFLOW_PLUGIN_VERSION,
             name: #connector_name,
-            create_vtable: mf_create_vtable,
+            create: mf_create,
         };
 
-        extern "C" fn mf_create_vtable() -> ::memflow::plugins::ConnectorFunctionTable {
-            ::memflow::plugins::ConnectorFunctionTable::create_vtable::<#connector_type>(mf_create)
+        #create_gen
+
+        #func
+    };
+
+    gen.into()
+}
+
+#[proc_macro_attribute]
+pub fn os_layer(args: TokenStream, input: TokenStream) -> TokenStream {
+    let attr_args = parse_macro_input!(args as AttributeArgs);
+    let args = match OSFactoryArgs::from_list(&attr_args) {
+        Ok(v) => v,
+        Err(e) => return TokenStream::from(e.write_errors()),
+    };
+
+    let os_name = args.name;
+
+    let func = parse_macro_input!(input as ItemFn);
+    let func_name = &func.sig.ident;
+
+    let create_gen = if func.sig.inputs.len() > 2 {
+        quote! {
+            #[doc(hidden)]
+            extern "C" fn mf_create(
+                args: ::memflow::types::ReprCStr,
+                mem: ::memflow::plugins::COption<::memflow::plugins::ConnectorInstance>,
+                log_level: i32,
+                out: &mut ::memflow::plugins::os::MUOSInstance
+            ) -> i32 {
+                ::memflow::plugins::os::create_with_logging(args, mem.into(), log_level, out, #func_name)
+            }
+        }
+    } else {
+        quote! {
+            #[doc(hidden)]
+            extern "C" fn mf_create(
+                args: ::memflow::types::ReprCStr,
+                mem: ::memflow::plugins::COption<::memflow::plugins::ConnectorInstance>,
+                _: i32,
+                out: &mut ::memflow::plugins::os::MUOSInstance
+            ) -> i32 {
+                ::memflow::plugins::os::create_without_logging(args, mem.into(), out, #func_name)
+            }
         }
+    };
+
+    let gen = quote! {
+        #[doc(hidden)]
+        #[no_mangle]
+        pub static MEMFLOW_OS: ::memflow::plugins::OSLayerDescriptor = ::memflow::plugins::OSLayerDescriptor {
+            os_version: ::memflow::plugins::MEMFLOW_PLUGIN_VERSION,
+            name: #os_name,
+            create: mf_create,
+        };
+
+        #create_gen
 
         #func
     };
 
-    gen.extend(create_gen);
+    gen.into()
+}
+
+#[proc_macro_attribute]
+pub fn os_layer_bare(args: TokenStream, input: TokenStream) -> TokenStream {
```

Comment:
The proc macros here should have at least an example in their documentation to show a potential user how to add them.